programs/svm-spoke/src/utils/message_utils.rs

```diff
@@ -0,0 +1,102 @@
+
+    // First remaining account is the handler and the rest are accounts to be passed to the message handler.
+    let message_accounts_len = message.accounts.len();
+    if remaining_accounts.len() != message_accounts_len + 1 {
+        return err!(AcrossPlusError::InvalidMessageKeyLength);
+    }
+    if (message.read_only_len as usize) > message_accounts_len {
+        return err!(AcrossPlusError::InvalidReadOnlyKeyLength);
+    }
+    let handler = &remaining_accounts[0];
+    let account_infos = &remaining_accounts[1..];
+
+    if handler.key() != message.handler {
+        return err!(AcrossPlusError::InvalidMessageHandler);
+    }
+
+    // Populate accounts for the invoked message handler CPI.
+    let mut accounts = Vec::with_capacity(message_accounts_len);
+    for (i, message_account_key) in message.accounts.into_iter().enumerate() {
+        if account_infos[i].key() != message_account_key {
+            return Err(Error::from(AcrossPlusError::InvalidMessageAccountKey)
+                .with_pubkeys((account_infos[i].key(), message_account_key)));
+        }
+
+        // Writable accounts must be passed first. This enforces the same write permissions as set in the message. Note
+        // that this would fail if any of mutable FillV3Relay / ExecuteV3SlowRelayLeaf accounts are passed as read-only
+        // in the bridged message as the calling client deduplicates the accounts and applies maximum required
+        // privileges. Though it is unlikely that any practical application would require this.
+        // We also explicitly disable all signer privileges for all the accounts to protect the relayer from being
+        // drained of funds in the inner instructions.
+        match i < message_accounts_len - (message.read_only_len as usize) {
+            true => {
+                if !account_infos[i].is_writable {
+                    return Err(Error::from(AcrossPlusError::NotWritableMessageAccountKey)
+                        .with_account_name(format!("{}", message_account_key)));
+                }
+                accounts.push(AccountMeta::new(message_account_key, false));
+            }
+            false => {
+                if account_infos[i].is_writable {
+                    return Err(Error::from(AcrossPlusError::NotReadOnlyMessageAccountKey)
+                        .with_account_name(format!("{}", message_account_key)));
+                }
+                accounts.push(AccountMeta::new_readonly(message_account_key, false));
+            }
+        }
+    }
+
+    // Transfer value amount from the relayer to the first account in the message accounts.
+    // Note that the depositor is responsible to make sure that after invoking the handler the recipient account will
+    // not hold any balance that is below its rent-exempt threshold, otherwise the fill would fail.
+    if message.value_amount > 0 {
```

Comment:
This is mostly useful when invoked handler transaction requires account initialization payer (e.g. when creating ATAs)